include/sampleflow/producers/delayed_rejection_mh.h

```diff
@@ -0,0 +1,164 @@
+     * sophisticated perturbation strategy.
+     */
+    template <typename OutputType>
+    class DelayedRejectionMetropolisHastings : public Producer<OutputType>
+    {
+      public:
+        /**
+         * The principal function of this class. Starting from the given
+         * initial sample $x_0$, it produces a sequence of samples $x_k$
+         * that are passed through the signal of the base class to
+         * Consumer objects.
+         *
+         * @param[in] starting_point The initial sample $x_0$.
+         * @param[in] log_likelihood A function object that, when called
+         *   with a sample $x$, returns $\log(\pi(x))$, i.e., the natural
+         *   logarithm of the likelihood function evaluated at the sample.
+         * @param[in] perturb A function object that, when given a sample
+         *   $x$ as well as a vector of rejected samples $y$,
+         *   returns a pair of values containing the following:
+         *   <ol>
+         *   <li> A different sample $\tilde x$ that is perturbed
+         *     in some way from the given sample $x$.
+         *   <li> The relative probability of the transition $x\to\tilde x$
+         *     divided by the probability of the transition $\tilde x\to x$.
+         *     Specifically, if the proposal distribution is given by
+         *     $\pi_\text{proposal}(\tilde x|x)$, then the second element
+         *     of the pair returned by the `perturb` argument is
+         *     $\frac{\pi_\text{proposal}(\tilde x|x)}
+         *           {\pi_\text{proposal}(x|\tilde x)}$.
+         *   </ol>
+         *   If samples are from some continuous space, say
+         *   ${\mathbb R}^n$, then the perturbation function is often
+         *   implemented by choosing $\tilde x$ from a neighborhood
+         *   of $x$. If, for example, $\tilde x$ is chosen with a probability
+         *   that only depends on the distance $\|\tilde x-\x\|$ as is often
+         *   done (e.g., uniformly in a disk of a certain radius around
+         *   $x$, or using a Gaussian probability distribution centered at
+         *   $x$), then the ratio returned as second argument is
+         *   $\frac{\pi_\text{proposal}(\tilde x|x)}
+         *           {\pi_\text{proposal}(x|\tilde x)}=1$.
+         * @param[in] n_samples The number of (new) samples to be produced
+         *   by this function. This is also the number of times the
+         *   signal is called that notifies Consumer objects that a new
+         *   sample is available.
+         * @param[in] max_delays The maximum number of delayed rejection stages.
+         */
+        void
+        sample (const OutputType &starting_point,
+                const std::function<double (const OutputType &)> &log_likelihood,
+                const std::function<std::pair<OutputType,double> (const OutputType &, const std::vector<OutputType>)> &perturb,
+                const types::sample_index n_samples,
+                const int max_delays);
```

Comment:
I would move this argument one up. (And do the same with the documentation.) That's because in some sense, the `log_likelihood` and `perturb` arguments along with `max_delays` are the inputs to each iteration of the algorithm, and should be listed together; `n_samples` falls into a different "category" of inputs.

Also, make it `unsigned int` -- it makes no sense to pass a negative number of delays.